![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Pablo Badilla, Diego Cortez
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Javiera Arévalo, Tamara Carrasco, Ignacio Reyes

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Bryan Cabezas 
- Estudiante n°2: Gonzalo Sobarzo


_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

El proyecto debe realizarse al final del curso y su objetivo es poner a prueba el conocimiento de todas las unidades del curso. Este consta de 3 partes: **Informe de Análisis** (30%), **Construcción del Modelo** (40%) y **Despliegue del Modelo** (30%). Este notebook corresponde a la **Parte 1 - Informe de Análisis**.

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. 
- **Plazo de entrega**: Hasta el Viernes 10 de Julio
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

## 📊 Distribución de Puntaje

Esta parte se evalúa en **escala de 1.0 a 7.0**, donde la nota se calcula como:

$$\text{Nota} = 1.0 + \text{puntaje obtenido}$$

El puntaje total repartible es de **6.0 puntos**, distribuidos así:

| Sección | Detalle | Puntaje |
| ------- | ------- | ------: |
| **Sección 1 - Entendimiento del Problema** | Preguntas conceptuales (P1–P7) | 2.0 |
| **Sección 2 · Subsección 1 - Datos Tabulares** | Código (1.0) + Preguntas (1.6) | 2.6 |
| **Sección 2 · Subsección 2 - Embeddings** | Código (0.6) + Preguntas (0.8) | 1.4 |
| **Total** | | **6.0** |

Cada pregunta y tarea indica su puntaje entre paréntesis, p. ej. _(0.3 pts)_.


Esta nota corresponde a la nota de la **Parte 1** que equivale a un **30%** de la nota del proyecto final. Para obtener la nota final, las notas individuales de cada parte se ponderan de la siguiente manera.

| Parte              | Ponderación |
|--------------------|------------:|
| **1 — Análisis**   | **30%**     |
| 2 — Modelo         | 40%         |
| 3 — Despliegue     | 30%         |

---

## 📖 Introducción 

<p align="center">
    <img src="./chaucherapp.jpg" alt="Alt Text" width="700">
</p>

La app chilena **ChaucherApp** ha emergido recientemente como una de las nuevas apps *FinTech* que apuesta por dominar en un competitivo mercado. Esta app ofrece una billetera digital que permite hacer transferencias, tarjetas digitales que permiten comprar, y servicios financieros como préstamos, seguros e inversiones.

Sin duda es sorprendente que **ChaucherApp** sea uno de los líderes de la industria fintech, siendo una empresa tan joven. En las palabras del CEO de la empresa, el Sr. Querty _'El gran éxito de ChaucherApp se debe a que entendimos la importancia de escuchar y responder eficientemente al cliente'_. En efecto, **ChaucherApp** fue diseñada desde el comienzo con una robusta arquitectura digital relacionada al **Soporte al cliente**.

Durante años, **ChaucherApp** ha logrado dar respuesta satisfactoria a cientos de tickets de soporte al cliente al mes, lo que le ha permitido inspirar confianza y prestigio. Por supuesto, esto también ha generado una gran cantidad de datos que pueden ser útiles para las siguientes etapas de crecimiento de la empresa. Y es que, a medida que el mercado se hace más competitivo, se hace más necesario aumentar la eficiencia de los procesos de respuesta de tickets, por lo que el Sr. Querty ha optado por **automatizar el proceso** incorporando inteligencia artificial.

Luego de intensa planificación ejecutiva y jornadas de analítica, han decidido que el primer paso para automatizar será el de desarrollar un modelo de **priorización de tickets** de soporte al cliente utilizando **modelos de lenguaje** y **aprendizaje de máquinas**. De esta manera podrán optimizar recursos humanos mientras se desarrollan otras herramientas.

Para lograr esta tarea, te han encargado analizar datos históricos de tickets de soporte al cliente que ya han sido manualmente priorizados, presentes en el archivo `data/tickets.parquet`. En él, encontrarás tickets con el contenido del ticket, categoría, canal, fecha y priorización, que puede ser `Baja`, `Media`, `Alta` o `Crítica`. Tu objetivo es utilizar esa data y los medios, algoritmos y estrategias que estimes convenientes para construir un **modelo clasificador** que sea capaz de predecir esta categoría en base al texto y atributos del ticket.

Con esta información, tus tareas a realizar son primero **analizar y entender los datos**, luego **construir un modelo de inferencia**, y finalmente **implementar ese modelo** para que el equipo de ejecutivos de ChaucherApp pueda utilizarlo.

---

# Sección 1 - Entendimiento del Problema

_(Total: 2.0 pts)_

El objetivo de esta sección es entender cuál será el problema que tendrás que resolver, incluyendo tanto la definición misma del problema (i.e., si es clasificación binaria, clasificación multiclase, regresión, clustering, otros...) como también las métricas, tanto de negocio como de evaluación, que usarás para medir el éxito de tu solución.

Para esto, primero responde las siguientes preguntas:


1. ¿Qué tipo de problema vamos a resolver? ¿Clasificación? ¿Regresión? ¿Clustering? ¿Reducción de dimensionalidad? ¿Ranking? ¿Recomendación? ¿Por qué? (utiliza la definición de cada uno de estos tipos de problemas más el enunciado para fundamentar). _(0.3 pts)_
2. ¿Por qué el negocio te solicitaría resolver este problema? ¿Por qué es importante? _(0.2 pts)_
3. ¿Por qué el negocio solicita generar un modelo predictivo para resolver este problema? ¿Por qué no lo pueden hacer, por ejemplo, por medio de código más palabras claves? _(0.2 pts)_
4. ¿Cuáles serían las métricas de evaluación relevantes para el negocio que buscarás optimizar? Aquí tienes la libertad de fundamentar lo que se te ocurra en base a tu entendimiento del negocio. Nombra al menos 3 y justifica la elección. _(0.3 pts)_
5. ¿Cuáles serían las métricas de evaluación de ML importantes para este problema? Para responder esto, primero da la interpretación de qué significa cada métrica (R^2, RMSE, MSE, etc.) / (Accuracy, Precision, Recall, F1) respecto a un problema binario Correcto / Incorrecto. Importante: Debes responder con la interpretación de qué significa cada una de estas métricas respecto al problema binario. _(0.5 pts)_
6. Luego, a partir de las interpretaciones anteriores de métricas más el conocimiento del negocio, elige cuál métrica es la más apropiada para cada clase en base a las prioridades del negocio. _(0.3 pts)_
7. Por último, ¿qué es la agregación macro? ¿Y la weighted? Y, ¿qué tipo de agregación utilizarías para promediar estas métricas de evaluación? _(0.2 pts)_

**Responde aquí**

> 
1- El problema a resolver es una clasifiacción multiclase que busca predecir la priorización de un ticket (ya sea alto, medio o bajo). Esto se realizará para automatizar el proceso, generando un aumento en la eficiencia de los procesos de respuesta de ticket, permitiendo optimizar los recursos humanos mientras se desarrollan nueva herramientas por parte de ChaucherApp. (FALTA DEFINICION DE CADA UNO DE LOS TIPOS)

2- Es importante porque se dispone de recursos humanos limitados, resultando que existan tickets con mayor rapidez de atención y solución que otros.

3- El negocio solicita generar un modelo predictivo porque el contenido del ticket puede variar en cada uno de ellos, generando que reglas simples no puedan abarcar la naturaleza de todos los problemas. Esto da puerta a que modelos de clasificación que puedan buscar relaciones internas y semantica del texto, lo cual con simples reglas presentaria limitaciones.

4- Una métrica de evaluación relevante para el negocio es maximizar los tickets respondidos al mes. Otra métrica es optimizar la cantidad de tickes criticos solucionados y tiempo de respuesta. (FALTA JUSTIFICAR)

5- La interpretación de las métricas respecto al problema binario de clasificación es: 
- Exactitud (Accuracy): Mide el total de predicciones correctas divididas por el total de datos. Es la métrica más rápida, pero puede ser engañosa si algunas clases tienen muchos más datos que otras.
- Precisión (Precision): La proporción de predicciones correctas para una clase específica respecto a todas las veces que el modelo predijo esa clase.
- Exhaustividad (Recall): La proporción de predicciones correctas para una clase respecto al total real de elementos que pertenecen a esa clase.
- Puntuación F₁ (F1-Score): Combina la precisión y la exhaustividad en una sola medida. Es muy útil si necesitas un equilibrio entre ambas métricas
Por el lado de regresión seria:
- R^2
- RMSE
- MAE (FALTA)
- MSE

Finalmente, para este problema la métrica más importante es el Recall, ya que buscamos minizar los falsos negativos (que clasifique un ticket con categoría baja cuando debería ser crítico). Adicionalmente, nuestra segunda métrica clave también es el F1-Score, ya que busca encontrar el equilibrio entre falsos positivos y falsos negativos.

6- Para las clases Crítico y Alto, se prioriza la métrica Recall, ya que buscamos maximizar la captura de casos graves. Para las clases media y baja sería precisión, lo cual mide la certeza de las alertas generadas, buscando asegurar que la anomalías sean precisas en estas clases.

7- La agregación macro es una métrica global que calcula el promedio simple de las métricas de cada clase, tratando a todos por igual. Mientras que, la agregación weighted calcula la métrica para cada clase y la multiplica por la proporción de instancias (tickets) que pertenencen a cada prioridad, generando un peso que refleja el desbalance de los datos. La agregación que utilizaríamos sería la macro, ya que buscamos un alta sensibilidad en la clase crítica, castigando al modelo en caso de fallar en una clase, dando mismo peso en promedio a todas las clases. De lo contrario, una agregación Weighted puede sesgar los resultados al presentar un desbalance entre clases.



---

# Sección 2 - Análisis de Datos

## Subsección 1 - Datos Tabulares

_(Total: 2.6 pts — código 1.0 + preguntas 1.6)_


La idea de esta segunda sección es que leas las fuentes de datos y las explores.


Partiremos primero con `tickets.parquet`. Utiliza la función de I/O asociada a este tipo de archivo para cargar los datos.

In [17]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots

In [2]:
# Tu código aquí...
df = pd.read_parquet('data/tickets.parquet')
df

,Id_Ticket,Asunto_Ticket,Contenido_Ticket,N_Caracteres_Ticket,Canal_Ticket,Categoría_Problema,Usuario-Tipo_de_Cuenta,Usuario-Antiguedad_Cuenta_Dias,Fecha_Envío,Tiempo_de_Resolución_Horas,Nivel_Prioridad
0,TCK-19267,Error al iniciar sesión en nuevo celular por d...,hola oye lo que pasa es que me cambie de celu ...,531,Whatsapp,Cuenta,Premium,159,2024-01-02,14,Baja
1,TCK-19268,Error al actualizar correo electrónico: bloque...,"Hola, buenas tardes. Escribo porque estoy trat...",784,Whatsapp,Otro,Premium,192,2024-01-02,25,Media
2,TCK-19269,Cobro desconocido en mi tarjeta digital,"Hola buenas, escribo porque me aparecio un des...",600,Whatsapp,Fraude,Free,314,2024-01-02,10,Baja
3,TCK-19270,"Cuenta bloqueada sin aviso, necesito ayuda urg...","Hola, necesito que me solucionen esto luego po...",659,Whatsapp,Cuenta,Free,228,2024-01-02,2,Media
4,TCK-19271,Me descontaron plata del cajero y no salió el ...,Hola buenos dias escribo porque ya estoy aburr...,853,Whatsapp,Otro,Free,72,2024-01-03,24,Media
...,...,...,...,...,...,...,...,...,...,...,...
1995,TCK-21262,Error al intentar ver datos de la tarjeta virtual,"Hola, queria consultarles una cuestion. Estaba...",615,Whatsapp,Pregunta general,Premium,43,2025-12-30,3,Media
1996,TCK-21263,No llega el código para recuperar mi contraseñ...,"Oiga, chiquillos de ChaucherApp, de verdad que...",754,Whatsapp,Pregunta general,Free,92,2025-12-31,2,Baja
1997,TCK-21264,Consulta sobre cambio de color o fondo en la app,"Hola buenas, oiga lo que pasa es que tengo una...",499,Página Web,Otro,Free,49,2025-12-31,19,Baja
1998,TCK-21265,Error al pagar cuenta: fondos insuficientes a ...,Oigan la wea molesta que me acaba de pasar... ...,700,Correo,Pregunta general,Free,54,2025-12-31,10,Media


Ahora, analiza los datos de entrada usando las herramientas solicitadas _(código: 1.0 pts en total)_:


1. Utiliza `.info()` y `.describe()` para obtener una visualización rápida del dataframe. _(0.1 pts)_
2. Utiliza `histogram` de plotly express para explorar la distribución de la variable objetivo, y `value_counts` más operaciones entre series para calcular la cantidad y porcentaje de sus categorías. _(0.2 pts)_
3. Utiliza `histogram` más [subplots](https://plotly.com/python/subplots/) para observar la distribución de características numéricas y categóricas. Hint: puedes acceder con `fig.data[0]` a la traza de un histograma generado por `histogram`. _(0.2 pts)_
4. Utiliza `corr` + `imshow` para analizar la correlación entre variables numéricas. _(0.1 pts)_
5. Genera bins con `qcut` para `N_Caracteres_Ticket`, `Tiempo_de_Resolución_Horas` y `Usuario-Antiguedad_Cuenta_Dias`, con los valores `Bajo`, `Medio`, `Alto`. Luego, utiliza `crosstab` (normalizando usando `normalize` por filas) + `imshow` para analizar las contingencias entre `Categoría_Problema` y `Nivel_Prioridad`, `Canal_Ticket` y `Nivel_Prioridad`, `Usuario-Tipo_de_Cuenta` y `Nivel_Prioridad`, y los bins de `N_Caracteres_Ticket` y `Nivel_Prioridad`, `Tiempo_de_Resolución_Horas` y `Nivel_Prioridad`, y `Usuario-Antiguedad_Cuenta_Dias` y `Nivel_Prioridad`. Grafica en subplots. _(0.2 pts)_
6. Por último, agrega la cantidad de reclamos por mes y `Nivel_Prioridad` (crea una nueva columna llamada `Fecha_Envío_Mes` usando el módulo de datetimes de pandas) para obtener un conteo de cantidad de reclamos según mes-nivel de prioridad, y utiliza `line` para generar un gráfico de línea en donde cada `Nivel_Prioridad` tenga su propia serie. _(0.2 pts)_

In [6]:
# Tu código aquí...
# Recuerda que puedes usar múltiples celdas.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                          Non-Null Count  Dtype
---  ------                          --------------  -----
 0   Id_Ticket                       2000 non-null   str  
 1   Asunto_Ticket                   2000 non-null   str  
 2   Contenido_Ticket                2000 non-null   str  
 3   N_Caracteres_Ticket             2000 non-null   int64
 4   Canal_Ticket                    2000 non-null   str  
 5   Categoría_Problema              2000 non-null   str  
 6   Usuario-Tipo_de_Cuenta          2000 non-null   str  
 7   Usuario-Antiguedad_Cuenta_Dias  2000 non-null   int64
 8   Fecha_Envío                     2000 non-null   str  
 9   Tiempo_de_Resolución_Horas      2000 non-null   int64
 10  Nivel_Prioridad                 2000 non-null   str  
dtypes: int64(3), str(8)
memory usage: 1.6 MB


In [7]:
df.describe()

,N_Caracteres_Ticket,Usuario-Antiguedad_Cuenta_Dias,Tiempo_de_Resolución_Horas
count,2000.000000,2000.000000,2000.000000
mean,717.699000,121.743000,18.718000
std,209.355096,142.339366,18.841788
min,251.000000,4.000000,1.000000
25%,564.000000,42.000000,5.000000
50%,697.000000,78.000000,13.000000
75%,850.250000,143.000000,26.000000
max,1831.000000,1200.000000,183.000000


In [12]:
orden_prioridades = ["Baja", "Media", "Alta", "Critica"]

plt.figure(figsize=(10, 6))
px.histogram(df, x='Nivel_Prioridad', title='Distribución de prioridades en los tickets', category_orders={'Nivel_Prioridad': orden_prioridades}, height=600).show()

<Figure size 1000x600 with 0 Axes>

In [14]:

cantidades = df["Nivel_Prioridad"].value_counts()
porcentajes = df["Nivel_Prioridad"].value_counts(normalize=True) * 100

resumen_prioridades = pd.DataFrame(
    {"Cantidad": cantidades, "Porcentaje (%)": porcentajes.round(2)}
)

resumen_prioridades

,Cantidad,Porcentaje (%)
Nivel_Prioridad,,
Media,789,39.45
Baja,723,36.15
Alta,350,17.50
Critica,138,6.90


In [53]:
import numpy as np
import plotly.express as px
from plotly.subplots import make_subplots

# 1. Definimos explícitamente cuáles variables queremos analizar
# Excluimos IDs, fechas y texto libre para que los histogramas tengan sentido
columnas_numericas = [
    "N_Caracteres_Ticket",
    "Usuario-Antiguedad_Cuenta_Dias",
    "Tiempo_de_Resolución_Horas",
]

columnas_categoricas = [
    "Canal_Ticket",
    "Categoría_Problema",
    "Usuario-Tipo_de_Cuenta",
    "Nivel_Prioridad",
]

columnas_totales = columnas_numericas + columnas_categoricas

# 2. Calculamos las dimensiones de la grilla de forma dinámica
total_variables = len(columnas_totales)
columnas_grilla = 2
filas_grilla = int(np.ceil(total_variables / columnas_grilla))

# 3. Creamos la grilla con los títulos correspondientes
fig_subplots = make_subplots(
    rows=filas_grilla,
    cols=columnas_grilla,
    subplot_titles=[f"Dist. de {col}" for col in columnas_totales],
)

# 4. Iteramos usando enumerate para no dejar ninguna columna en el camino
for idx, col in enumerate(columnas_totales):
    fila = (idx // columnas_grilla) + 1
    columna = (idx % columnas_grilla) + 1

    # Personalizamos el orden si es la columna de Nivel_Prioridad
    if col == "Nivel_Prioridad":
        fig_temporal = px.histogram(
            df, x=col, category_orders={col: ["Baja", "Media", "Alto", "Crítico"]}
        )
    elif col == "Tiempo_de_Resolución_Horas":
        # Usamos log_y=True para que las frecuencias bajas no desaparezcan
        fig_temporal = px.histogram(df, x=col, log_y=True)
    else:
        fig_temporal = px.histogram(df, x=col)

    # Extraemos la traza con fig.data[0] tal como lo pide el Hint
    fig_subplots.add_trace(fig_temporal.data[0], row=fila, col=columna)

# 5. Desvinculamos los ejes para que cada variable mantenga su escala e independencia
fig_subplots.update_xaxes(matches=None)
fig_subplots.update_yaxes(matches=None)

# 6. Ajustamos el tamaño total para que las 4 filas de gráficos se vean perfectas
fig_subplots.update_layout(
    height=filas_grilla * 320,
    width=950,
    title_text="<b>Análisis Exploratorio: Distribución de Características de Tickets</b>",
    showlegend=False,
)

fig_subplots.show()

In [34]:
variables_numericas = [
    "N_Caracteres_Ticket",
    "Usuario-Antiguedad_Cuenta_Dias",
    "Tiempo_de_Resolución_Horas",
]
df_numerico = df[variables_numericas]

matriz_corr = df_numerico.corr()

# 3. Graficamos el mapa de calor utilizando imshow
fig_corr = px.imshow(
    matriz_corr,
    text_auto=".2f",  # Muestra los valores con 2 decimales dentro de cada cuadro
    aspect="auto",
    color_continuous_scale="RdBu_r",  # Escala de color clásica (Rojo = Positiva, Azul = Negativa)
    labels=dict(color="Correlación"),
    x=matriz_corr.columns,
    y=matriz_corr.columns,
    title="<b>Matriz de Correlación entre Variables Numéricas</b>",
)

# Ajustamos tamaño para que sea perfectamente legible
fig_corr.update_layout(width=600, height=500)

fig_corr.show()

In [ ]:

etiquetas = ["Bajo", "Medio", "Alto"]


df["N_Caracteres_Ticket_Bin"] = pd.qcut(
    df["N_Caracteres_Ticket"], q=3, labels=etiquetas, duplicates="drop"
)
df["Tiempo_de_Resolución_Horas_Bin"] = pd.qcut(
    df["Tiempo_de_Resolución_Horas"], q=3, labels=etiquetas, duplicates="drop"
)
df["Usuario-Antiguedad_Cuenta_Dias_Bin"] = pd.qcut(
    df["Usuario-Antiguedad_Cuenta_Dias"], q=3, labels=etiquetas, duplicates="drop"
)


variables_analisis = [
    "Categoría_Problema",
    "Canal_Ticket",
    "Usuario-Tipo_de_Cuenta",
    "N_Caracteres_Ticket_Bin",
    "Tiempo_de_Resolución_Horas_Bin",
    "Usuario-Antiguedad_Cuenta_Dias_Bin",
]

orden_prioridades = ["Baja", "Media", "Alto", "Crítico"]


fig_subplots = make_subplots(
    rows=3,
    cols=2,
    subplot_titles=[f"Contingencia: {var} vs Prioridad" for var in variables_analisis],
)


for idx, var in enumerate(variables_analisis):
    fila = (idx // 2) + 1
    columna = (idx % 2) + 1

    tabla_contingencia = pd.crosstab(
        df[var], df["Nivel_Prioridad"], normalize="index"
    )
    columns_existentes = [
        p for p in orden_prioridades if p in tabla_contingencia.columns
    ]
    tabla_contingencia = tabla_contingencia[columns_existentes]

    fig_temporal = px.imshow(
        tabla_contingencia,
        text_auto=".2f",
        color_continuous_scale="YlOrRd",
    )


    fig_subplots.add_trace(fig_temporal.data[0], row=fila, col=columna)


    fig_subplots.update_xaxes(
        tickmode="array",
        tickvals=list(range(len(columns_existentes))),
        ticktext=columns_existentes,
        row=fila,
        col=columna,
    )
    fig_subplots.update_yaxes(
        tickmode="array",
        tickvals=list(range(len(tabla_contingencia.index))),
        ticktext=[str(i) for i in tabla_contingencia.index],
        row=fila,
        col=columna,
    )


fig_subplots.update_layout(
    height=1000,
    width=950,
    title_text="<b>Análisis de Contingencias Normalizadas por Fila (vs Nivel_Prioridad)</b>",
    coloraxis={"colorscale": "YlOrRd"},
)

fig_subplots.show()

In [46]:
df['Fecha_Envío'] = df['Fecha_Envío'].astype('datetime64[ns]')
df['Fecha_Envio_Mes'] = df['Fecha_Envío'].dt.month

df_conteo_mensual = (
    df.groupby(['Fecha_Envio_Mes', 'Nivel_Prioridad']).size().reset_index(name='Cantidad_reclamos_mes')
)


orden_prioridades = ["Baja", "Media", "Alta", "Critica"]

fig_lineas = px.line(
    df_conteo_mensual,
    x="Fecha_Envio_Mes",
    y="Cantidad_reclamos_mes",
    color="Nivel_Prioridad",  # Crea una serie/línea por cada nivel de prioridad
    category_orders={"Nivel_Prioridad": orden_prioridades},
    title="<b>Evolución Mensual de Reclamos por Nivel de Prioridad</b>",
    labels={
        "Fecha_Envio_Mes": "Mes de Envío",
        "Cantidad_reclamos_mes": "Cantidad de Reclamos",
        "Nivel_Prioridad": "Prioridad",
    },
    markers=True,  # Agrega puntitos en cada mes para que sea más fácil leer los valores precisos
)

# 4. Ajustes de diseño
fig_lineas.update_layout(
    width=900,
    height=500,
    xaxis_tickangle=-45,  # Inclina los meses para que no se encimen en el eje X
)

fig_lineas.show()

Ahora, responde las siguientes preguntas _(preguntas: 1.6 pts en total)_:

1. ¿Cuántos datos son? ¿Hay nulos? _(0.1 pts)_
2. ¿Qué columnas tenemos disponibles? ¿Qué tipo de dato es cada columna? - Tanto en tipo de dato (str, int, float, etc.) como también en su contenido (texto libre, categorías, entero, número decimal, fechas, índices) y su relación. Responde en una tabla con: Columna, Tipo de dato, Tipo de contenido y Descripción. _(0.2 pts)_
3. ¿Cómo se distribuyen las etiquetas? ¿Hay desbalance? Incluye porcentajes en la respuesta. _(0.2 pts)_
4. ¿Cómo afecta la distribución de clases a las respuestas dadas en la sección 1? ¿Sería necesario editar o corregir algo de estas? Si es así, edita las respuestas de la sección 1 y menciona qué edición hiciste aquí. _(0.1 pts)_
5. ¿Cómo se distribuyen las variables numéricas? ¿Con cuáles distribuciones las podríamos relacionar? ¿Cuál es la interpretación de estas? ¿Existen outliers? Explica por cada variable numérica. _(0.2 pts)_
6. ¿Existe alguna correlación entre las variables numéricas? _(0.1 pts)_
7. ¿Existe alguna relación interesante entre las categorías generadas y las categorías de la variable objetivo? _(0.2 pts)_
8. ¿Existe alguna relación entre el tiempo y la variable objetivo? ¿Sería recomendable utilizar esta variable? _(0.1 pts)_
9. Pensemos ahora por un momento en el modelo en modo inferencia/producción. ¿Existe alguna variable que no vaya a estar disponible en ese momento? Si es así, ¿cómo se llama este fenómeno? ¿Si no hubiésemos meditado esto, qué hubiese sucedido? _(0.2 pts)_
10. Para generar los bins de las variables numéricas se utilizó `qcut` (cuantiles) en lugar de `cut` (ancho fijo). ¿Qué habría pasado con el heatmap de contingencia de `Tiempo_de_Resolución_Horas` si hubieras usado `cut`? ¿Por qué `qcut` es preferible cuando la variable está muy sesgada? _(0.2 pts)_

In [47]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 15 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Id_Ticket                           2000 non-null   str           
 1   Asunto_Ticket                       2000 non-null   str           
 2   Contenido_Ticket                    2000 non-null   str           
 3   N_Caracteres_Ticket                 2000 non-null   int64         
 4   Canal_Ticket                        2000 non-null   str           
 5   Categoría_Problema                  2000 non-null   str           
 6   Usuario-Tipo_de_Cuenta              2000 non-null   str           
 7   Usuario-Antiguedad_Cuenta_Dias      2000 non-null   int64         
 8   Fecha_Envío                         2000 non-null   datetime64[ns]
 9   Tiempo_de_Resolución_Horas          2000 non-null   int64         
 10  Nivel_Prioridad                    

**Responde aquí**

> 

1- Presentamos 2000 tickets con 0 nulos
2- Las columnas disponibles son:
| Columna | Tipo de Dato | Tipo de Contenido | Descripción |
| :--- | :--- | :--- | :--- |
| **Id_Ticket** | `str` (String) | Identificador Único | Código único de registro asignado a cada ticket de soporte. |
| **Asunto_Ticket** | `str` (String) | Texto Libre Corto | Resumen o título descriptivo del problema reportado por el usuario. |
| **Contenido_Ticket** | `str` (String) | Texto Libre Largo | Cuerpo del mensaje detallado con el relato completo o reclamo enviado por el usuario. |
| **N_Caracteres_Ticket** | `int64` (Entero) | Numérico Continuo | Cantidad total de caracteres presentes en el texto del cuerpo del ticket. |
| **Canal_Ticket** | `str` (String) | Categórico Nominal | Medio a través del cual se recibió el reclamo. |
| **Categoría_Problema** | `str` (String) | Categórico Nominal | Clasificación temática del incidente. |
| **Usuario-Tipo_de_Cuenta** | `str` (String) | Categórico Nominal | Plan de suscripción que posee el cliente en la plataforma. |
| **Usuario-Antiguedad_Cuenta_Dias** | `int64` (Entero) | Numérico Continuo | Tiempo transcurrido en días desde la apertura de la cuenta del usuario hasta la fecha del ticket. |
| **Fecha_Envío** | `str` / `datetime64` | Temporal / Fecha | Registro cronológico del día en que el usuario envió el ticket de reclamo (Formato `AAAA-MM-DD`). |
| **Tiempo_de_Resolución_Horas** | `int64` (Entero) | Numérico Continuo | Cantidad de horas que tomó el equipo de soporte en resolver y cerrar el caso. |
| **Nivel_Prioridad** | `str` (String) | Categórico Ordinal | Severidad o urgencia asignada al ticket para su atención (Valores: `Baja`, `Media`, `Alto`, `Crítico`). |

3- Si existe desbalance en las clases del nivel de prioridad. Por un lado, la clase baja presenta 723 valores con un 36.15% de la muestra. La clase media presenta 789 valores siendo el 39.45% de los datos. La clase alta únicamente preesnta 350 datos (17.50%) y finalmente la clase critica solo 138 datos representando un 6.90% de la muestra siendo la clase minoritaria.

4- La distribución de los datos no afectan directamente a la respuesta de la sección anterior, ya que el desbalanceo de clases era lo esperado. Dado esto, nuestra métrica clave sigue siendo el recall, es decir priorizamos la clase minoritaria.

5- La variable numérica N_Caracteres_ticket se distribuye muy similar a una distribución normal, pero sesgada hacia la izquierda, con una cola larga hacia los caracteres altos. Es decir, en su mayoría los tickets presentan entre 500 y 1000 caracteres, pero existe una minoria de tickets outliers con el número de caracteres mayor a 1.500. La variable "Usuario_antiguedad_cuenta_dias" presenta una distribución con sesgo negativo hacia valores bajos, representando que la mayoria de los usuarios presentan una antiguedad baja, probablemente al comenzar a utilizar los servicios de la empresa. Existen valores outliers en una minoria de usuaios que llegan a más de 1000 días de antiguedad. La varaible "Tiempo de resolución horas" presenta nuevamente una distribución sesgada negativamente hacia valores bajos, demostrando que existe una alta eficiencia en la respuesta por parte del área de soporte pero aún con ciertos outliers en tiempo mayores a 150 días, incluso con un ticket de más de 1.500 días.

6- No existe correlación entre variable númericas, siendo el máximo 0.19. 

7- Las categorías generadas mediante qcut demuestran una relación clara y un fuerte poder discriminante frente a la variable objetivo. Se observa que los tickets cortos (N_Caracteres_Ticket_Bin en Bajo) se concentran masivamente en prioridad Baja (68%), mientras que al volverse más extensos (Medio) migran hacia una prioridad Media (51%), lo que indica que los usuarios detallan más el texto ante incidentes complejos. Asimismo se revela que la antigüedad del cliente influye en la severidad, ya que los usuarios nuevos registran menores tasas de urgencia frente a los clientes de Alta antigüedad, quienes elevan sus reclamos a prioridad Media en un 45%. En contraste, los bins de Tiempo_de_Resolución_Horas_Bin muestran una distribución homogénea y plana entre el 30% y 43% en todas sus etapas, sugiriendo que la resolución operativa en estos niveles no está condicionada exclusivamente por la etiqueta de prioridad inicial, consolidando a las métricas de texto y perfil del usuario como los mejores predictores para segmentar el riesgo o la urgencia de los reclamos.

8- Sí, existe una relación directa y visible entre el tiempo y la variable objetivo. Se observan fluctuaciones estacionales cruzadas muy marcadas, como en los meses 6, 8 y 9, donde el volumen de reclamos de prioridad Baja y Media se comporta de forma inversa, alternando picos de saturación en el servicio. Este comportamiento estacional afecta directamente la carga operativa y los tiempos de respuesta del equipo, por lo que sí es altamente recomendable utilizar esta variable. Para integrarla con éxito en el modelo, no se debe usar la fecha cruda, sino transformarla mediante ingeniería de características en componentes cíclicos (como el mes o el día de la semana) o en variables de rezago (lags) que capturen el historial reciente de saturación, permitiendo predecir y segmentar los picos de urgencia de manera óptima.


9- En producción, las variables que no tendremos a disposición serían: Tiempo de resolución horas y el nivel de prioridad. Por otro lado, la varaible Categoría problema puede existir o no, dependiendo si el usuario al ingresar la solicitud agrega la categoría respectiva, de lo contrario seria una variable faltante. Este fenomeno se llama Data Leakage y pasaría que el modelo haría una predicción perfecta si se le entrega esta información.

10- Si hubieras utilizado cut en lugar de qcut, el heatmap de contingencia de Tiempo_de_Resolución_Horas se habría vuelto ilegible y vacío en su mayoría. Debido al fuerte sesgo a la izquierda, donde la gran masa de tickets se concentra en las primeras horas, mientras que unos pocos outliers se extienden hasta rangos muy altos. Aplicar cut habría agrupado casi el 100% de los datos en la categoría "Bajo", dejando los bins de "Medio" y "Alto"con frecuencias insignificantes. Por esta razón, qcut es preferible cuando una variable está muy sesgada, ya que divide los intervalos basándose en la cantidad de registros (cuantiles) y no en la distancia del rango.

---

## Subsección 2 - Embeddings

_(Total: 1.4 pts — código 0.6 + preguntas 0.8)_

En esta sección trabajarás con embeddings sobre los Asuntos + Descripciones de tickets precalculados con el modelo `gemini-embeddings-001`. Ahora, nuevamente utilizando las herramientas de lectura de parquet _(código: 0.6 pts en total)_:

1. Lee el archivo que contiene los embeddings (`data/embeddings.parquet`). _(0.1 pts)_
2. Reduce la dimensionalidad de los embeddings con umap-learn a dos dimensiones y agrega a `df_tickets` las nuevas dimensiones. Asegura tu proyección utilizando semillas, así luego, al momento de la revisión, se podrán replicar los resultados. _(0.3 pts)_
3. Grafica un scatter plot con las proyecciones y pinta por `Nivel_Prioridad` (utiliza el código dado).
4. Grafica un nuevo scatter plot con las proyecciones y pinta por `Categoría_Problema`. _(0.2 pts)_

In [ ]:
# Tu código aquí...
# Recuerda que puedes usar múltiples celdas.



In [ ]:
import textwrap


def wrap_text(text, width=40):
    return "<br>".join(textwrap.wrap(text, width=width))


df_tickets["Contenido_Ticket_Wrapped"] = df_tickets["Contenido_Ticket"].apply(lambda x: wrap_text(x, 80))

fig_proj = px.scatter(
    df_tickets,
    x="x",
    y="y",
    color="Nivel_Prioridad",
    title="Proyecciones de los embeddings respecto a clase",
    hover_data=[
        "Canal_Ticket",
        "N_Caracteres_Ticket",
        "Categoría_Problema",
        "Usuario-Tipo_de_Cuenta",
        "Usuario-Antiguedad_Cuenta_Dias",
        "Contenido_Ticket_Wrapped",
    ],
)
fig_proj.show()

Ahora, responde las siguientes preguntas _(preguntas: 0.8 pts en total)_:

1. ¿Qué es un embedding? ¿Qué representan? A grandes rasgos, ¿cómo se calculan? Hint: consulta sobre `gemini-embedding-001`. _(0.2 pts)_
2. ¿Por qué los embeddings tienen el potencial de ser mejores para un modelo predictivo que analizar las palabras individuales? _(0.1 pts)_
3. ¿Cuál podría ser una desventaja de utilizar embeddings en vez de utilizar directamente palabras? _(0.1 pts)_
4. ¿Existen otros tipos de embeddings? ¿Qué podrían representar? _(0.1 pts)_
5. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Nivel_Prioridad`? _(0.1 pts)_
6. ¿Qué patrón interesante puede mostrar el gráfico de proyecciones pintadas por `Categoría_Problema`? _(0.1 pts)_
7. Las proyecciones de UMAP (`x`, `y`) quedaron agregadas a `df_tickets`. ¿Sería buena idea usar estas 2 dimensiones como features de entrada del modelo predictivo, en lugar de los embeddings originales? Justifica pensando en la información que se pierde al reducir a 2D. _(0.1 pts)_

**Responde aquí**

> ...